great_expectations/expectations/expectation.py

```diff
@@ -330,6 +330,7 @@ class Expectation(pydantic.BaseModel, metaclass=MetaExpectation):
 
     class Config:
         arbitrary_types_allowed = True
+        smart_union = True
```

Comment:
Default for Pydantic V2 so let's use it here - iterate through all possible types before coercing